### ISSUE TITLE
Feature request: output full HTTP request and response in a JSON structure

### ISSUE THREAD

**Issue by hughpv** [httpie/httpie#1007]
I would love to capture the full HTTP request and response in a JSON structure so I can extract the various parts programmatically without parsing the plaintext output.

For example:

```
http --output-full-json ...

{
  "request": {
    "headers": "Accept: application/json\r\nContent-Type: application/json",
    "body": "<request body here, escaped/encoded as necessary to work as a string value>"
  },
  "response": {
    "headers": "201 Created\r\nContent-Type: application/json\r\nContent-Length: 42",
    "body": "<response body here, escaped/encoded as necessary to work as a string value>"
  }
}
```

**Comment by jakubroztocil** [comment]
Thanks for the suggestion, @hughpv. I’ve been considering this for a while.

#### Initial thoughts

* [Output options](https://httpie.io/docs#output-options) should be respected.
* With [`--all`](https://httpie.io/docs#showing-intermediary-redirect-responses), an `Array` of exchanges is printed instead of a single object.
* What about binary bodies? Apply base64 encoding like [httpbin.org](http://httpbin.org/)? Let user choose to skip/base64-encode?
* A structured object for headers would be nice, but we should support multiple headers of the same name somehow.
* UI — `--output-format` which could be `raw` (default) + `json` + potentially more in the future.
* User can always pipe to `jq` but consider supporting simple selectors directly. Something like `--select=response.headers.content-type`. These would work on the JSON structure so when used, then `--output-format=json` would be implied.

#### Hypothetical example

```bash
$ http -v --output-format=json  example.org
```

```jsonc
// Exchange
{
  "request": {
    "headers": {
      "Foo": "Bar"
    },
    "body": "…"
  },
  "response": {
    "headers": {
      "Foo": "Bar"
    },
    "body": "…"
  }
}
```

**Comment by hughpv** [comment]
Awesome ideas; love them all. Let me just expand a little on my use case --

I have it in mind to build a test suite that, in addition to testing functionality, also captures output for use in documentation, e.g.:

```asciidoc
= Example Request
----
include:request.http[]
----

= Example Response
----
include:response.http[]
----
```

... where `request.http` and `response.http` would be built from the httpie output.

Thus what I'm looking for is essentially a "verbatim" or "raw" option. All the other stuff makes sense but the raw output was the main idea I came here with.

Thanks!

**Comment by jakubroztocil** [comment]
I see, so for your use case, the structured headers would be a complication, because you would have to manually _un-structure_ them for the docs?

**Comment by hughpv** [comment]
Correct. :) But I could definitely see where some users would find structured headers quite useful.

**Comment by Ousret** [comment]
Hi,

I might have something to help with headers. https://github.com/Ousret/kiss-headers 
Should help with a potential JSON repr.

**Comment by ahmadnassri** [comment]
[HAR](https://en.wikipedia.org/wiki/HAR_(file_format)) would be a great format to use for this

**Comment by morgen-peschke** [comment]
This would be _extremely_ helpful for some testing use cases (I've currently got a really ugly shell function that mimics this with vanilla `curl`).

Some thoughts on the format:
* It's probably worth exposing the HTTP code as a number (makes success checks really easy)
* Wrapping the header values in an array may provide a low-friction way to deal with multiple array values.
  The default case only requires adding `[0]`, which is a pretty low cost.
* It may be worth treating it as an opt-in sort of thing, as this format is more likely to be useful in scripts than interactively, so being a little more explicit is probably not going to be an issue.

Borrowing language from `git`, this could be an alternative to `--print`: `--porcelain`.

### Examples
#### Raw
Command:
```
--porcelain=req.headers.raw,req.body.text,req.method,req.uri.raw,resp.headers.raw,resp.body.text,resp.code,resp.status
```
Output:
```
{
  "request": {
   "uri": "http://example.com:8081/test?foo=bar",
   "method": "POST",
    "headers": "Accept: application/json\r\nContent-Type: application/json",
    "body": "Lorem ipsum dolor sit amet"
  },
  "response": {
    "code": 201,
    "status": "Created",
    "headers": "Content-Type: application/json\r\nContent-Length: 42",
    "body": "consectetur adipiscing elit"
  }
}
```
#### Structured/Parsed
Command:
```
--porcelain=req.headers.json,req.body.base64,req.method,req.uri.json,resp.headers.raw,resp.body.text,resp.code,resp.status
```
Output:
```
{
  "request": {
   "uri": { 
     "scheme": "http", 
     "host": "example.com", 
     "port": 8081,
     "path": "/test",
     "query": {
       "foo": ["bar"]
   "method": "POST",
    "headers": {
      "Accept": ["application/json"],
	  "Content-Type": [application/json]"
    },
    "body": "TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQ="
  },
  "response": {
    "code": 201,
    "status": "Created",
    "headers": "Content-Type: application/json\r\nContent-Length: 42",
    "body": "consectetur adipiscing elit"
  }
}
```
